tests/suites/test_suite_ssl.data

```diff
@@ -272,6 +272,58 @@ Handshake, ECDH-ECDSA-WITH-CAMELLIA-256-CBC-SHA384
 depends_on:MBEDTLS_SHA384_C:MBEDTLS_CIPHER_MODE_CBC:MBEDTLS_ECDSA_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED:MBEDTLS_ECP_DP_SECP384R1_ENABLED:MBEDTLS_CAMELLIA_C
 handshake_cipher:"TLS-ECDH-ECDSA-WITH-CAMELLIA-256-CBC-SHA384":MBEDTLS_PK_ECDSA:0
 
+Handshake, select ECDHE-RSA-WITH-AES-256-GCM-SHA384, non-opaque
+depends_on:MBEDTLS_SHA384_C:MBEDTLS_AES_C:MBEDTLS_GCM_C:MBEDTLS_RSA_C:MBEDTLS_ECP_DP_SECP384R1_ENABLED:MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED:MBEDTLS_SSL_PROTO_DTLS
+handshake_ciphersuite_select:"TLS-ECDHE-RSA-WITH-AES-256-GCM-SHA384":MBEDTLS_PK_RSA:PSA_ALG_NONE:PSA_ALG_NONE:0:0:0:MBEDTLS_TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+
+Handshake, select ECDHE-RSA-WITH-AES-256-GCM-SHA384, opaque
+depends_on:MBEDTLS_SHA384_C:MBEDTLS_AES_C:MBEDTLS_GCM_C:MBEDTLS_RSA_C:MBEDTLS_ECP_DP_SECP384R1_ENABLED:MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED:MBEDTLS_SSL_PROTO_DTLS:MBEDTLS_USE_PSA_CRYPTO
+handshake_ciphersuite_select:"TLS-ECDHE-RSA-WITH-AES-256-GCM-SHA384":MBEDTLS_PK_RSA:PSA_ALG_RSA_PKCS1V15_SIGN( PSA_ALG_ANY_HASH ):PSA_ALG_NONE:PSA_KEY_USAGE_SIGN_HASH:0:0:MBEDTLS_TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+
+Handshake, select ECDHE-RSA-WITH-AES-256-GCM-SHA384, opaque, invalid alg
+depends_on:MBEDTLS_SHA384_C:MBEDTLS_AES_C:MBEDTLS_GCM_C:MBEDTLS_RSA_C:MBEDTLS_ECP_DP_SECP384R1_ENABLED:MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED:MBEDTLS_SSL_PROTO_DTLS:MBEDTLS_USE_PSA_CRYPTO
+handshake_ciphersuite_select:"TLS-ECDHE-RSA-WITH-AES-256-GCM-SHA384":MBEDTLS_PK_RSA:PSA_ALG_RSA_PKCS1V15_SIGN( PSA_ALG_SHA_256 ):PSA_ALG_NONE:PSA_KEY_USAGE_SIGN_HASH:0:MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE:0
```

Comment:
I tried to add `PSA_ALG_SHA_384` in thue particular positive `Handshake, select ECDHE-RSA-WITH-AES-256-GCM-SHA384, opaque` test but it fails because the picked hash for ServerKeyExchange is SHA512.

It's selected in `ssl_parse_sig_alg_ext()`, and since SHA512 is the first one for ECDSA, it's kept without checking if the key supports it.

Probably `ssl->handshake->sig_algs` should only contain hash algorithms supported by the key.

Reply:
The problem is that all this hash selection is done after ciphersuite & cert is selected, at some point sig_algs should be stripped down with only hash algs supported by the certificate key.

Reply:
The following code makes it work:

```patch
diff --git a/library/ssl_tls12_server.c b/library/ssl_tls12_server.c
index 05fa5ab6b..c79089402 100644
--- a/library/ssl_tls12_server.c
+++ b/library/ssl_tls12_server.c
@@ -1125,6 +1125,47 @@ static int ssl_ciphersuite_match( mbedtls_ssl_context *ssl, int suite_id,
     }
 #endif
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    if( ssl->handshake->key_cert &&
+        ssl->handshake->key_cert->key &&
+        ssl->handshake->sig_algs )
+    {
+        const uint16_t *p = ssl->handshake->sig_algs;
+
+        /* Reset hash_algs */
+        mbedtls_ssl_sig_hash_set_init( &ssl->handshake->hash_algs );
+
+        /* Loop over sig_algs and add the compatible ones to has_algs set */
+        for( ; *p != MBEDTLS_TLS1_3_SIG_NONE; ++p )
+        {
+            mbedtls_md_type_t md_type = MBEDTLS_BYTE_1( *p );
+            mbedtls_pk_type_t pk_type;
+            psa_algorithm_t alg;
+
+            if( ( MBEDTLS_BYTE_0( *p ) ) == MBEDTLS_SSL_SIG_ECDSA )
+            {
+                pk_type = MBEDTLS_PK_ECDSA;
+                alg = PSA_ALG_ECDSA( mbedtls_psa_translate_md( md_type ) );
+            }
+            else if( ( MBEDTLS_BYTE_0( *p ) ) == MBEDTLS_SSL_SIG_RSA )
+            {
+                pk_type = MBEDTLS_PK_RSA;
+                alg = PSA_ALG_RSA_PKCS1V15_SIGN(
+                                    mbedtls_psa_translate_md( md_type ) );
+            }
+            else
+                continue;
+
+            if( mbedtls_pk_can_do_ext( ssl->handshake->key_cert->key,
+                                         alg, PSA_KEY_USAGE_SIGN_HASH ) )
+            {
+                mbedtls_ssl_sig_hash_set_add( &ssl->handshake->hash_algs,
+                                              pk_type, md_type );
+            }
+        }
+    }
+#endif
+
     *ciphersuite_info = suite_info;
     return( 0 );
 }
```

After `ssl_pick_cert()`, it runs throught the `sig_algs` and only calls `mbedtls_ssl_sig_hash_set_add()` if the key supports the alg + hash.

With this, the picked hash from `hash_algs` is the right one.

Reply:
Oh, good point. I realise now I was confused: I thought the signature hash was always the same as the hash that's named in the ciphersuite, as in `AES-128-GCM-SHA384`) but it's not the case, this hash is negotiated separately. So indeed, this patch looks like the correct approach.

I'm glad I asked for a positive test case then :)

Reply:
I wonder how we should handle if the only hash alg supported by the key is not in the `sig_alg_ext` buffer ?

In `ssl_parse_sig_alg_ext` we should store the intersection of the offered algs and the supported algs in `sig_alg` and then use it to select the ciphersuite working with the key, no ?

Reply:
Yes, we want to compute the intersection and if it's empty we should not select that ciphersuite-key pair (which might either result in another being selected, or an early handshake failure).

Reply:
If this an acceptable way to solve this I'll commit it:
```diff
diff --git a/library/ssl_misc.h b/library/ssl_misc.h
index c2c3cd210..884f66fde 100644
--- a/library/ssl_misc.h
+++ b/library/ssl_misc.h
@@ -602,8 +602,7 @@ struct mbedtls_ssl_handshake_params
     mbedtls_ssl_sig_hash_set_t hash_algs;             /*!<  Set of suitable sig-hash pairs */
 #endif
 
-#if defined(MBEDTLS_SSL_PROTO_TLS1_3) && \
-    defined(MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED)
+#if defined(MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED)
     uint16_t received_sig_algs[MBEDTLS_RECEIVED_SIG_ALGS_SIZE];
 #endif
 
diff --git a/library/ssl_tls12_server.c b/library/ssl_tls12_server.c
index 05fa5ab6b..168d4380f 100644
--- a/library/ssl_tls12_server.c
+++ b/library/ssl_tls12_server.c
@@ -228,6 +228,13 @@ static int ssl_parse_sig_alg_ext( mbedtls_ssl_context *ssl,
                                   size_t len )
 {
     size_t sig_alg_list_size;
+    uint16_t sig_alg;
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    uint32_t common_idx = 0;
+
+    memset( ssl->handshake->received_sig_algs, 0,
+            sizeof(ssl->handshake->received_sig_algs) );
+#endif
 
     const unsigned char *p;
     const unsigned char *end = buf + len;
@@ -251,15 +258,6 @@ static int ssl_parse_sig_alg_ext( mbedtls_ssl_context *ssl,
         return( MBEDTLS_ERR_SSL_DECODE_ERROR );
     }
 
-    /* Currently we only guarantee signing the ServerKeyExchange message according
-     * to the constraints specified in this extension (see above), so it suffices
-     * to remember only one suitable hash for each possible signature algorithm.
-     *
-     * This will change when we also consider certificate signatures,
-     * in which case we will need to remember the whole signature-hash
-     * pair list from the extension.
-     */
-
     for( p = buf + 2; p < end; p += 2 )
     {
         /* Silently ignore unknown signature or hash algorithms. */
@@ -280,9 +278,17 @@ static int ssl_parse_sig_alg_ext( mbedtls_ssl_context *ssl,
             continue;
         }
 
-        if( mbedtls_ssl_sig_alg_is_offered(
-                ssl, MBEDTLS_GET_UINT16_BE( p, 0 ) ) )
+        sig_alg = MBEDTLS_GET_UINT16_BE( p, 0 );
+
+        if( mbedtls_ssl_sig_alg_is_offered( ssl, sig_alg ) )
         {
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+            if( common_idx + 1 < MBEDTLS_RECEIVED_SIG_ALGS_SIZE )
+            {
+                ssl->handshake->received_sig_algs[common_idx] = sig_alg;
+                common_idx += 1;
+            }
+#endif
             mbedtls_ssl_sig_hash_set_add( &ssl->handshake->hash_algs, sig_cur, md_cur );
             MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello v3, signature_algorithm ext:"
                                         " match sig %u and hash %u",
@@ -295,6 +301,18 @@ static int ssl_parse_sig_alg_ext( mbedtls_ssl_context *ssl,
         }
     }
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    if( common_idx == 0 )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "no signature algorithm in common" ) );
+        mbedtls_ssl_send_alert_message( ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL,
+                                    MBEDTLS_SSL_ALERT_MSG_HANDSHAKE_FAILURE );
+        return( MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE );
+    }
+
+    ssl->handshake->received_sig_algs[common_idx] = MBEDTLS_TLS1_3_SIG_NONE;
+#endif
+
     return( 0 );
 }
 #endif /* MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED */
@@ -1095,20 +1113,6 @@ static int ssl_ciphersuite_match( mbedtls_ssl_context *ssl, int suite_id,
     }
 #endif
 
-#if defined(MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED)
-    /* If the ciphersuite requires signing, check whether
-     * a suitable hash algorithm is present. */
-    sig_type = mbedtls_ssl_get_ciphersuite_sig_alg( suite_info );
-    if( sig_type != MBEDTLS_PK_NONE &&
-        mbedtls_ssl_sig_hash_set_find( &ssl->handshake->hash_algs, sig_type ) == MBEDTLS_MD_NONE )
-    {
-        MBEDTLS_SSL_DEBUG_MSG( 3, ( "ciphersuite mismatch: no suitable hash algorithm "
-                                    "for signature algorithm %u", (unsigned) sig_type ) );
-        return( 0 );
-    }
-
-#endif /* MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED */
-
 #if defined(MBEDTLS_X509_CRT_PARSE_C)
     /*
      * Final check: if ciphersuite requires us to have a
@@ -1125,6 +1129,74 @@ static int ssl_ciphersuite_match( mbedtls_ssl_context *ssl, int suite_id,
     }
 #endif
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    /*
+     * Now the certificate/key is selected, test each signature/hash algorithms
+     * supported for this hanshake and only memorize the first hash algorithm
+     * of each signature algorithm supported by the key.
+     */
+    if( ssl->handshake->key_cert &&
+        ssl->handshake->key_cert->key )
+    {
+        uint32_t alg_idx = 0;
+
+        /* Reset hash_algs */
+        mbedtls_ssl_sig_hash_set_init( &ssl->handshake->hash_algs );
+
+        /* Loop over received_sig_algs and add the compatible ones to has_algs set */
+        while( alg_idx < MBEDTLS_RECEIVED_SIG_ALGS_SIZE )
+        {
+            uint16_t sig_alg = ssl->handshake->received_sig_algs[alg_idx];
+            mbedtls_md_type_t md_type = MBEDTLS_MD_NONE;
+            mbedtls_pk_type_t pk_type = MBEDTLS_PK_NONE;
+            psa_algorithm_t alg = PSA_ALG_NONE;
+
+            if( sig_alg == MBEDTLS_TLS1_3_SIG_NONE )
+                break;
+
+            md_type = MBEDTLS_BYTE_1( sig_alg );
+
+            if( ( MBEDTLS_BYTE_0( sig_alg ) ) == MBEDTLS_SSL_SIG_ECDSA )
+            {
+                pk_type = MBEDTLS_PK_ECDSA;
+                alg = PSA_ALG_ECDSA( mbedtls_psa_translate_md( md_type ) );
+            }
+            else if( ( MBEDTLS_BYTE_0( sig_alg ) ) == MBEDTLS_SSL_SIG_RSA )
+            {
+                pk_type = MBEDTLS_PK_RSA;
+                alg = PSA_ALG_RSA_PKCS1V15_SIGN(
+                                    mbedtls_psa_translate_md( md_type ) );
+            }
+
+            if( pk_type != MBEDTLS_PK_NONE &&
+                mbedtls_pk_can_do_ext( ssl->handshake->key_cert->key,
+                                       alg, PSA_KEY_USAGE_SIGN_HASH ) )
+            {
+                mbedtls_ssl_sig_hash_set_add( &ssl->handshake->hash_algs,
+                                              pk_type, md_type );
+                MBEDTLS_SSL_DEBUG_MSG( 3, ( "hash algs add:"
+                                        " sig %u and hash %u",
+                                        (unsigned) pk_type, (unsigned) md_type ) );
+            }
+
+            alg_idx += 1;
+        }
+    }
+#endif
+
+#if defined(MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED)
+    /* If the ciphersuite requires signing, check whether
+     * a suitable hash algorithm is present. */
+    sig_type = mbedtls_ssl_get_ciphersuite_sig_alg( suite_info );
+    if( sig_type != MBEDTLS_PK_NONE &&
+        mbedtls_ssl_sig_hash_set_find( &ssl->handshake->hash_algs, sig_type ) == MBEDTLS_MD_NONE )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "ciphersuite mismatch: no suitable hash algorithm "
+                                    "for signature algorithm %u", (unsigned) sig_type ) );
+        return( 0 );
+    }
+#endif /* MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED */
+
     *ciphersuite_info = suite_info;
     return( 0 );
 }
```

Reply:
Forget, most of the same work was done in #5785, with this merge I only had to add the check with  `mbedtls_pk_can_do_ext` in `mbedtls_ssl_tls12_get_preferred_hash_for_sig_alg()`

Reply:
Oh, cool, thanks for noticing!